crates/walrus-core/src/encoding/quilt_encoding.rs

```diff
@@ -0,0 +1,1468 @@
+        add_blob_to_data(&final_index_data, 0);
+
+        tracing::debug!("construct quilt success {}", data.len());
+
+        Ok(QuiltV1 {
+            data,
+            row_size,
+            quilt_index,
+            symbol_size,
+        })
+    }
+
+    /// Encodes the blobs into a quilt and returns the slivers.
+    fn encode(&self) -> Result<Vec<SliverPair>, QuiltError> {
+        let _guard = self.span.enter();
+        tracing::debug!("starting to encode quilt");
+
+        let quilt = self.construct_quilt()?;
+        let encoder = BlobEncoder::new(self.config.clone(), quilt.data()).map_err(|_| {
+            QuiltError::QuiltOversize(format!("quilt is too large: {}", quilt.data().len()))
+        })?;
+        assert_eq!(encoder.symbol_usize(), quilt.symbol_size());
+        Ok(encoder.encode())
+    }
+
+    /// Encodes the blobs into a quilt and returns the slivers and metadata.
+    fn encode_with_metadata(&self) -> Result<(Vec<SliverPair>, QuiltMetadata), QuiltError> {
+        let _guard = self.span.enter();
+        tracing::debug!("starting to encode quilt with metadata");
+
+        let quilt = self.construct_quilt()?;
+        let encoder = BlobEncoder::new(self.config.clone(), quilt.data()).map_err(|_| {
+            QuiltError::QuiltOversize(format!("quilt is too large: {}", quilt.data.len()))
+        })?;
+
+        assert_eq!(encoder.symbol_usize(), quilt.symbol_size);
+
+        let (sliver_pairs, metadata) = encoder.encode_with_metadata();
+        let quilt_metadata = QuiltMetadata::V1(QuiltMetadataV1 {
+            quilt_blob_id: *metadata.blob_id(),
+            metadata: metadata.metadata().clone(),
+            index: QuiltIndexV1 {
+                quilt_patches: quilt.quilt_index().quilt_patches.clone(),
+            },
+        });
+
+        Ok((sliver_pairs, quilt_metadata))
+    }
+}
+
+/// A quilt decoder of version V1.
+#[derive(Debug)]
+pub struct QuiltDecoderV1<'a> {
+    slivers: HashMap<SliverIndex, &'a SliverData<Secondary>>,
+    quilt_index: Option<QuiltIndexV1>,
+}
+
+impl<'a> QuiltDecoderApi<'a, QuiltVersionV1> for QuiltDecoderV1<'a> {
+    fn get_or_decode_quilt_index(&mut self) -> Result<&QuiltIndexV1, QuiltError> {
+        if self.quilt_index.is_some() {
+            return Ok(self.quilt_index.as_ref().expect("quilt index should exist"));
+        }
+
+        let first_sliver_index = SliverIndex(0);
+        let first_sliver = self
+            .slivers
+            .get(&first_sliver_index)
+            .ok_or(QuiltError::MissingSlivers(vec![first_sliver_index]))?;
+
+        utils::check_quilt_version::<QuiltVersionV1>(first_sliver.symbols.data())?;
+
+        let data_size = utils::get_quilt_index_data_size(first_sliver.symbols.data())?;
+
+        // Calculate how many slivers we need based on the data size.
+        let num_slivers_needed = data_size.div_ceil(first_sliver.symbols.data().len());
+        let prefix_size = QUILT_INDEX_SIZE_PREFIX_SIZE + QUILT_TYPE_SIZE;
```

Comment:
Nit: This is used in 5 places; how about defining a separate constant `PREFIX_SIZE`?